src/librustc_typeck/check/writeback.rs

```diff
@@ -90,24 +90,51 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
     // we observe that something like `a+b` is (known to be)
     // operating on scalars, we clear the overload.
     fn fix_scalar_binary_expr(&mut self, e: &hir::Expr) {
-        if let hir::ExprBinary(ref op, ref lhs, ref rhs) = e.node {
-            let lhs_ty = self.fcx.node_ty(lhs.id);
-            let lhs_ty = self.fcx.infcx().resolve_type_vars_if_possible(&lhs_ty);
-
-            let rhs_ty = self.fcx.node_ty(rhs.id);
-            let rhs_ty = self.fcx.infcx().resolve_type_vars_if_possible(&rhs_ty);
-
-            if lhs_ty.is_scalar() && rhs_ty.is_scalar() {
-                self.fcx.inh.tables.borrow_mut().method_map.remove(&MethodCall::expr(e.id));
-
-                // weird but true: the by-ref binops put an
-                // adjustment on the lhs but not the rhs; the
-                // adjustment for rhs is kind of baked into the
-                // system.
-                if !hir_util::is_by_value_binop(op.node) {
-                    self.fcx.inh.tables.borrow_mut().adjustments.remove(&lhs.id);
+        match e.node {
+            hir::ExprBinary(ref op, ref lhs, ref rhs) |
+                hir::ExprAssignOp(ref op, ref lhs, ref rhs) => {
+                    let lhs_ty = self.fcx.node_ty(lhs.id);
+                    let lhs_ty = self.fcx.infcx().resolve_type_vars_if_possible(&lhs_ty);
+
+                    let rhs_ty = self.fcx.node_ty(rhs.id);
+                    let rhs_ty = self.fcx.infcx().resolve_type_vars_if_possible(&rhs_ty);
+
+                    if lhs_ty.is_scalar() && rhs_ty.is_scalar() {
+                        self.fcx.inh.tables.borrow_mut().method_map.remove(&MethodCall::expr(e.id));
+
+                        // weird but true: the by-ref binops put an
+                        // adjustment on the lhs but not the rhs; the
+                        // adjustment for rhs is kind of baked into the
+                        // system.
+                        match e.node {
+                            hir::ExprBinary(..) => {
+                                if !hir_util::is_by_value_binop(op.node) {
+                                    self.fcx.inh.tables.borrow_mut().adjustments.remove(&lhs.id);
+                                }
+                            },
+                            hir::ExprAssignOp(..) => {
+                                self.fcx.inh.tables.borrow_mut().adjustments.remove(&lhs.id);
+                            },
+                            _ => {},
+                        }
+                    } else {
+                        let tcx = self.tcx();
+
+                        if let hir::ExprAssignOp(..) = e.node {
+                            if !tcx.sess.features.borrow().augmented_assignments &&
+                                !self.fcx.expr_ty(e).references_error() {
```

Comment:
Nit: the body of the if and the condition ought not be aligned the same without an intervening `{`, imo. :) Can you move the opening brace to the next line, or else modify the indent, or something? I personally often do this but I think I am the only one that does so: ;)

```rust
if
    cond1 &&
    cond2
{
    ...
}
```